File: src/main/java/com/github/kornilova_l/flamegraph/plugin/server/trees/Filter.java
package com.github.kornilova_l.flamegraph.plugin.server.trees;

import com.github.kornilova_l.flamegraph.proto.TreeProtos.Tree.Node;
import org.jetbrains.annotations.Nullable;

import java.util.regex.Pattern;

public class Filter {
    @Nullable
    private final Pattern include;
    @Nullable
    private final Pattern exclude;

    public Filter(@Nullable String include, @Nullable String exclude) {
        this.include = compilePattern(include);
        this.exclude = compilePattern(exclude);
    }

    @Nullable
    private Pattern compilePattern(@Nullable String patternString) {
        if (patternString != null) {
            return Pattern.compile(
                    patternString.replaceAll("\\*", ".*")
                            .replaceAll("\\.", "\\.")
            );
        }
        return null;
    }

    public boolean isNodeIncluded(Node node) {
        String nodeString = getNodeString(node);
        boolean isIncluded = true;
        if (include != null) {
            isIncluded = include.matcher(nodeString).matches();
        }
        boolean isExcluded = false;
        if (exclude != null) {
            isExcluded = exclude.matcher(nodeString).matches();
        }
        return isIncluded &&
                !isExcluded;
    }

    private String getNodeString(Node node) {
        Node.NodeInfo nodeInfo = node.getNodeInfo();
        return nodeInfo.getClassName() + nodeInfo.getMethodName();
    }
}

File: src/main/java/com/github/korniloval/flameviewer/trees/Filter.kt
package com.github.korniloval.flameviewer.trees

import com.github.kornilova_l.flamegraph.proto.TreeProtos.Tree.Node
import com.intellij.openapi.diagnostic.Logger

import java.util.regex.Pattern
import java.util.regex.PatternSyntaxException

class Filter(include: String?, exclude: String?) {
    private val LOG = Logger.getInstance(Filter::class.java)
    private val include: Pattern?
    private val exclude: Pattern?

    init {
        this.include = compilePattern(include)
        this.exclude = compilePattern(exclude)
    }

    private fun compilePattern(patternString: String?): Pattern? {
        if (patternString != null) {
            return try {
                Pattern.compile(
                        patternString.replace(".", """\.""")
                                .replace("*", ".*")
                                .replace("$", """\$""")
                )
            } catch (e: PatternSyntaxException) { // if pattern is invalid
                LOG.error(e)
                null
            }
        }
        return null
    }

    fun isNodeIncluded(node: Node): Boolean {
        val nodeString = getNodeString(node)
        return isIncluded(nodeString) && !isExcluded(nodeString)
    }

    private fun isIncluded(nodeString: String): Boolean {
        if (include == null) {
            return true
        }
        return include.matcher(nodeString).matches()
    }

    private fun isExcluded(nodeString: String): Boolean {
        if (exclude == null) {
            return false
        }
        return exclude.matcher(nodeString).matches()
    }


    private fun getNodeString(node: Node): String {
        val nodeInfo = node.nodeInfo
        return nodeInfo.className + nodeInfo.methodName
    }
}
